paho-mqtt: add package and example (#12647)
RIOT-OS/RIOT: examples/paho-mqtt/README.md

### About
This application demonstrates the usage of the MQTT paho package in RIOT.

### Setup
For using this example, two prerequisites have to be fulfilled:

1. You need a running MQTT broker like Mosquitto broker for example. Take a look at
[Mosquitto Broker](https://mosquitto.org/). Check online any guide that will help you setting up the broker into some port (a).
For example this one for debian base linux users [How to setup a Mosquitto MQTT Server and receive data](https://www.digitalocean.com/community/questions/how-to-setup-a-mosquitto-mqtt-server-and-receive-data-from-owntracks).

2. Your RIOT node needs to be able to speak to that broker at the same port you set in 1.

### Setting up RIOT `native` on Linux

> - Run the script create.sh in dist/tools

Reviewer: Where is this script ? Better give the command directly:
```
sudo ./dist/tools/tapsetup/tapsetup -c 1
```

> - Replace the value of *DEFAULT_MQTT_BROKER_IP* with the IPv6 address provide by `ifconfig tapbr0`

Reviewer: I would drop this line and explain the `ifconfig tapbr0` in the `Running the example` section.
You can also drop the lines about port and the one after.

But I would just give an instruction for running the native instance:
```
make -C examples/paho-mqtt all term
```

- Change the default *MQTT_PORT* if needed
- You'll be able to change to broker IP or connection PORT within the `con` shell command if needed.
When running the example

## Running the example

- To connect to a broker, use the `con` command:
```

> con <broker ip> <broker port> <clientID> <username passwd> <keepalivetime>

Reviewer: This doesn't work with the current `con` command implementation. The usage should be:
```
con <broker ip> <broker port> <keepalivetime> <clientID> <username> <passwd>
```
(keepalivetime is not at the right place)

```

> * If no parameter is set, it will connect to *DEFAULT_MQTT_BROKER_IP*:*MQTT_PORT* with user=riot password=riot

Reviewer: I would not do it this way. con should be called with the explicit parameters given, no default for the broker address/port/clientId/username/password (in the Makefile or as constants in main.c)

Reviewer 2: Sorry to get into the discussion so late, but I have been completely busy in the last few months.

Now, where I'm trying to reproduce the problem with ESP32 boards, I'm wondering why all these parameters are mandatory. What if the broker does not require a user authentication like the Mosquito installation on my localhost? Why has the port to be specified if the MQTT standard port is used?

Author: Yes I agree with you after all

  * *clientID*: is the client id you set up on the broker.
  * *username and passwd*: those set in the broker, check online tutorial to do it regarding chosen broker.
  * *keepalivetime*: keep alive in second for your client. (Default one is 10 secs)

- To subscribe to a topic, run `sub` with the topic name as parameter and a QoS level between 1 to 3 (optional), e.g.
```

> sub hello/world [2]

Reviewer: What does `[2]` mean ? is it a reference ? If it's the QoS level, then just give an example, so it can be copy pasted from the readme (people are lazy).

### About
This application demonstrates the usage of the MQTT paho package in RIOT.

### Setup
For using this example, two prerequisites have to be fulfilled:

1. You need a running MQTT broker like Mosquitto broker for example. Take a look at
[Mosquitto Broker](https://mosquitto.org/). Check online any guide that will help you setting up the broker into some port (a).
For example this one for debian base linux users [How to setup a Mosquitto MQTT Server and receive data](https://www.digitalocean.com/community/questions/how-to-setup-a-mosquitto-mqtt-server-and-receive-data-from-owntracks).

2. Your RIOT node needs to be able to speak to that broker at the same port you set in 1.

### Setting up RIOT `native` on Linux
- Run `sudo ./dist/tools/tapsetup/tapsetup -c 1`

## Running the example
- Run on RIOT's root directory:

  make -C examples/paho-mqtt all term

- To connect to a broker, use the `con` command:
```

> con <ipv6 addr> <port> <KeepAliveInterval in sec> <user> <user ID> <password>

Reviewer: ```suggestion
con <ipv6 addr> <port> <KeepAliveInterval in sec> <client ID> <user> <password>
```

Author: `    data.username.cstring = argv[4];
    data.clientID.cstring = argv[5];
    data.password.cstring = argv[6];` regarding what I wrote, that are not the right positions, and I think client ID is not so important as username and password. I'll  s/user ID/client ID/ though

Author: Any ways, I think the order clientID user pass, it's good for a better undertandig

```
  * *ipv6 addr*: IPv6 broker address.
  * *port*: broker port.
  * *keepalivetime*: keep alive in second for your client.

>   * *user ID*: is the client id you set up on the broker.

Reviewer: ```suggestion
  * *client ID*: is the client id you set up on the broker.
```

### About
This application demonstrates the usage of the Eclipse paho MQTT package in RIOT.

### Setup
For using this example, two prerequisites have to be fulfilled:

1. You need a running MQTT broker like Mosquitto broker for example. Take a look at
[Mosquitto Broker](https://mosquitto.org/). Check online any guide that will
help you setting up the broker into some port (a).
For example this one for debian base linux users
[How to setup a Mosquitto MQTT Server and receive data](https://www.digitalocean.com/community/questions/how-to-setup-a-mosquitto-mqtt-server-and-receive-data-from-owntracks).

2. Your RIOT node needs to be able to speak to that broker at the same port you set in 1.

### Setting up RIOT `native` on Linux
- Run `sudo ./dist/tools/tapsetup/tapsetup -c 1`

## Running the example
- Run on RIOT's root directory:

  make -C examples/paho-mqtt all term

- To connect to a broker, use the `con` command:
```
con  <broker ip addr> [port] [clientID] [user] [password] [keepalivetime]
```
  * *broker ip addr*: IPv6 or IPv4 broker address.
  * *port*: broker port. Default 1883
  * *client ID*: is the client id you set up on the broker. Default can be set

>      through DEFAULT_MQTT_CLIENTID in your makefile. Otherwise is an empty string.

Reviewer: ```suggestion
     through DEFAULT_MQTT_CLIENT_ID in your makefile. Otherwise is an empty string.
```

Author: nice catch!

- To connect to a broker, use the `con` command:
```

> con <ipv6 addr> [user] [password] [port] [keepalivetime]

Reviewer 2: What about `Client ID`, it is described but not in the synopsis.

Author: I'm user the same as UserID

Author: I can add it but don't see it relevant. where you able to use it with IPv4?
